### ISSUE TITLE
All MUTATE and ALTER failed with "While proposing error: context deadline exceeded"

### ISSUE THREAD

**Issue by BlueT** [dgraph-io/dgraph#2492]
If you suspect this could be a bug, follow the template.

- What version of Dgraph are you using?

Docker image `dgraph/dgraph:latest` Image ID `f440f51e7171`

```
Dgraph version   : v1.0.6
Commit SHA-1     : ae8e5ffd
Commit timestamp : 2018-06-29 19:27:35 -0700
Branch           : HEAD
```

---
- Have you tried reproducing the issue with latest release?

Yes I'm using latest release.

---
- What is the hardware spec (RAM, OS)?

Hardware
```
CPU Intel(R) Core(TM) i5-4210U
Memory 16G, swap 16G
SSD Samsung 850 EVO 1T (~540MB/s R/W)
```
OS
```
Linux Zorya.BlueT.org 4.15.0-27-generic #29~16.04.1-Ubuntu SMP Wed Jul 11 10:03:45 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial
```
---

- Steps to reproduce the issue (command/config used to run Dgraph).

Run, feed data, count, restart, query, alter and mutate.

1. `docker-compose up`, where the docker-compose.yml is from docs.dgraph.io with little modifications.

2. Feed/insert millions of data.

Code was from the example script in official JavaScript client, with little modifications.
Batch insert. 5000 objects an array. Around 3,500,000 JSON object in total.

Schema:
```
name: string @index(hash) @lang .
id: string @index(exact) .
connection: uid @count @reverse .
```

Data: 3,000,000 records in `{id: "10 digit string", name: "UTF-8 Chinese string", type: "10 char string"}` and 30,000 Records in `{name: "UTF-8 Chinese string", type: "10 char string"}`.

3. Play on Dgraph Ratel Dashboard http://localhost:8000

Count nodes. `{ cn(func: has(name)) { count(uid) } }`.
**This Operation failed** but I saw dgraph still consume a lot CPU for minutes after dashboard shows operation failed.

4. `docker-compose down`.

5. Wait a while.

6. `docker-compose up`.

7. Query a node (success)

8. Try to ALTER schema on dashbaord.

**This Operation failed** with "Could not alter schema: While proposing error: context deadline exceeded"

```
type: string @index(exact) .
```

Console msg:
```
dgraph-server_1  | 2018/07/18 01:19:35 server.go:280: Got schema: [predicate:"type" value_type:STRING directive:INDEX tokenizer:"exact" ]
dgraph-zero_1    | 2018/07/18 01:19:38 oracle.go:76: purging below ts:50011, len(o.commits):0, len(o.aborts):0, len(o.rowCommit):0
```

9. Try to MUTATE add a node on dashboard.

**This Operation failed** with "While proposing error: context deadline exceeded"

```
{
  set {
  	_:lala <name> "lala" .
  	_:lala <type> "Natural" .
  }
}
```

Console msg:
```
dgraph-server_1  | 2018/07/18 01:20:53 server.go:280: Got schema: [predicate:"_share_hash_" value_type:STRING directive:INDEX tokenizer:"exact" ]
dgraph-zero_1    | 2018/07/18 01:20:58 oracle.go:76: purging below ts:50013, len(o.commits):0, len(o.aborts):0, len(o.rowCommit):0
dgraph-zero_1    | 2018/07/18 01:21:08 oracle.go:76: purging below ts:50014, len(o.commits):0, len(o.aborts):0, len(o.rowCommit):0
```

- Expected behavior and actual result.

Expected behavior: operation success.
Actual result: "While proposing error: context deadline exceeded"

Tried restarting dgraph multiple times, and tried to do the same operations multiple times, same results.

**Comment by manishrjain** [comment]
Did you try setting the schema before doing the load? I suspect, Dgraph is trying to build the indices from the existing data, which is why you're seeing a high CPU usage. Killing it won't help, because until those indices are built, Dgraph would keep trying.

**Comment by BlueT** [comment]
If needed, I can provide my dgraph docker volume.

Code was from the example script in official JavaScript client https://github.com/dgraph-io/dgraph-js/blob/master/examples/simple/index.js

```javascript
const dgraph = require("dgraph-js");
const grpc = require("grpc");
const csv = require('csvtojson');

const csvFilePath='natural.csv';
// const csvFilePath='juristic.csv';

const dgraphClientStub = newClientStub();
const dgraphClient = newClient(dgraphClientStub);

// const pagination = 100000;
const pagination = 5000;


// Create a client stub.
function newClientStub() {
	return new dgraph.DgraphClientStub("localhost:9080", grpc.credentials.createInsecure());
}

// Create a client.
function newClient(clientStub) {
	return new dgraph.DgraphClient(clientStub);
}


// Drop All - discard all data and start from a clean slate.
async function dropAll(dgraphClient) {
	const op = new dgraph.Operation();
	op.setDropAll(true);
	await dgraphClient.alter(op);
}

// Set schema.
async function setSchema(dgraphClient) {
	const schema = `
		name: string @index(hash) @lang .
		type: string @index(exact) .
		id: string @index(exact) .
		connection: uid @count @reverse .
	`;
	const op = new dgraph.Operation();
	op.setSchema(schema);
	await dgraphClient.alter(op);
}


// Create data using JSON.
async function createEntity(dgraphClient, entity) {
	// Create a new transaction.
	const txn = dgraphClient.newTxn();
	try {
				// Run mutation.
		const mu = new dgraph.Mutation();
		mu.setSetJson(entity);
		const assigned = await txn.mutate(mu);

		// Commit transaction.
		await txn.commit();

		// console.log("All created nodes (map from blank node names to uids):");
		// assigned.getUidsMap().forEach((uid, key) => console.log(`${key} => ${uid}`));
		// console.log("Total: " + assigned.getUidsMap().length);
		// console.log();
	} finally {
		// Clean up. Calling this after txn.commit() is a no-op
		// and hence safe.
		await txn.discard();
	}
}


// Query for data.
async function queryEntity(dgraphClient, {type, name}) {
	// Run query.
	const query = `query all($type: string, $name: string) {
	all(func: eq(name, $name)) @filter(eq(type, $type)) {
	    uid
	}
    }`;
	const vars = {
		$type: type,
		$name: name
	};
	const res = await dgraphClient.newTxn().queryWithVars(query, vars);
	const ppl = res.getJson();

	// Print results.
	console.log(`Number of people named "` + name + `": ${ppl.all.length}`);
	ppl.all.forEach((person) => console.log(person));
	return ppl.all[0];
}

async function main() {
	
	console.log("Starting at " + Date());
	await dropAll(dgraphClient);
	console.log("Existing data dropped at " + Date());
	await setSchema(dgraphClient);
	console.log("Schema updated at " + Date());

	console.log("Start reading CSV file at " + Date());
	return csv().fromFile(csvFilePath);
}


main()
.then(async (jsonObj) => {
	console.log(jsonObj);
	console.log("CSV to JSON loaded at " + Date());

	// do paging
	let pages = Math.floor(jsonObj.length / pagination);
	for (let i = 0; i <= pages; i++ ) {
		// console.log(jsonObj[i]);
		if (i === 0) {
			continue;
		}
		console.log("Page " + i + " (" + pagination + " records per page, feeding record " + (i-1)*pagination + " to " + i*pagination + ") at " + Date());
		await createEntity(dgraphClient, jsonObj.slice( (i-1)*pagination, i*pagination) );
	}

	// feed the rest records
	if (jsonObj.length > pages*pagination) {
		console.log("Feeding rest records (" + pagination + " records per page, feeding record " + pages*pagination + " to " + jsonObj.length + ") at " + Date());
		await createEntity(dgraphClient, jsonObj.slice( pages*pagination, jsonObj.length) );
	}

})
.then(() => {
	console.log("\nDONE! at " + Date());
	// Close the client stub.
	dgraphClientStub.close();
})
.catch((e) => {
	console.log("ERROR: ", e);
	console.log(Date());
});
```

**Comment by BlueT** [comment]
docker-compose.yml
```
version: "3.2"
services:
  dgraph-zero:
    image: dgraph/dgraph:latest
    volumes:
      - type: volume
        source: dgraph
        target: /dgraph
        volume:
          nocopy: true
    ports:
      - 5080:5080
      - 6080:6080
    restart: on-failure
    command: dgraph zero --my=dgraph-zero:5080 --expose_trace=true
  dgraph-server:
    image: dgraph/dgraph:latest
    volumes:
      - type: volume
        source: dgraph
        target: /dgraph
        volume:
          nocopy: true
    ports:
      - 8080:8080
      - 9080:9080
    depends_on:
      - dgraph-zero
    restart: on-failure
    command: dgraph server --my=dgraph-server:7080 --lru_mb=4096 --zero=dgraph-zero:5080 --expose_trace=true --badger.vlog=disk
  ratel:
    image: dgraph/dgraph:latest
    volumes:
      - type: volume
        source: dgraph
        target: /dgraph
        volume:
          nocopy: true
    ports:
      - 8000:8000
    depends_on:
      - dgraph-server
    command: dgraph-ratel

volumes:
  dgraph:
  
```

**Comment by BlueT** [comment]
@manishrjain thanks for quick reply. :-)
1. Yes I've set schema before loading data. But want to add one more index.
2. Actually dgraph did not even start comsuming CPU for these ALTER and MUTATE operations.
3. At step 3 (the `count()` part), CPU and Memory went high even after dashboard showed the operation was failed. But I had waited it went down before restarting containers.

**Comment by nishiltamboli** [comment]
Even I have been facing this issue on Mutate recently! Queries work as expected.

**Comment by manishrjain** [comment]
I've seen this issue. I have a PR to fix it. Will test it a bit more, and push out a binary with the fix, so you can try it out.

**Comment by manishrjain** [comment]
This commit should fix this issue. https://github.com/dgraph-io/dgraph/commit/8b0295b58115a01aec02d0a58c1fac56fe193235

**Comment by anne-markis** [comment]
Thanks @manishrjain 

I haven't had time to test this yet the problem hasn't happened to me again. What are the changes this will be included in 1.0.7?

**Comment by nmabhinandan** [comment]
I'm getting same error while mutating (with a large 32k line query) using web ui

**Comment by denggeng** [comment]
I have been facing this issue today. I known that the server has been restarted last night.

**Comment by danielmai** [comment]
@denggeng is the issue resolved for you in v1.0.7?

**Comment by manishrjain** [comment]
This issue should have been resolved with fixes in both v1.0.7 and v1.0.8. If it still exists, feel free to file a new issue.
